### PR TITLE
CONCF-967 | Remove check for duplicated title inside CC Desktop tool

### PR DESCRIPTION
**Pull request by vforge**
/cc @Warkot @tomasznapieralski @bognix 

Note:
- As a side of this ticket we switched from `title`-based error messages from server-side validation to `label`-based - this was followed with refactoring of desktop tool.

### CODE REVIEW

**Comment by tomasznapieralski** [comment]
LGTM

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js [hunk 1]
```diff
@@ -140,8 +142,8 @@ require(['wikia.window', 'jquery', 'wikia.nirvana', 'wikia.tracker', 'JSMessages
 				checkForm = function () {
 					$save.removeClass();
 
-					checkInputs($ul.find('.section-input'), ['limit', 'checkEmpty']);
-					checkInputs($ul.find('.item-input'), ['required', 'checkEmpty']);
+					checkInputs($ul.find('.section-input'), ['limit', 'checkEmpty', 'duplicates']);
```
Comment (bognix) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: do we need checkEmpty The only place where it's checked is inside if (checkEmpty || duplicate) and since they both go together maybe we can get rid of one.

Comment (vforge) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: Good suggestion - I'll test this assumption.

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js [hunk 2]
```diff
@@ -111,17 +112,18 @@ require(['wikia.window', 'jquery', 'wikia.nirvana', 'wikia.tracker', 'JSMessages
 							$this.addError(tooLongLabelError);
 							return true;
 						}
-						// check if value already exists (in cachedVals variable)
-						if (cachedVals.indexOf(val) === -1) {
-							// not exists, add it to cachedVals and remove previous errors
-							cachedVals.push(val);
-
-							$this.removeError();
```
Comment (rogatty) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js:123]: Is this change intended? We don't want to remove error when value doesn't exist in cache?

Comment (vforge) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js:123]: Yes, I've moved remove error to the bottom of the function to remove old (previous iteration) error states.

Comment (rogatty) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js:123]: But now you return from function before error is removed. Is that ok?

Comment (vforge) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js:123]: Return just goes for next `each` iteration, and error is added before each `return` statement. It's intended.

Comment (vforge) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js:123]: Also: it works ;)

**Comment by bognix** [comment]
@vforge - I can't save Curated Content on MLP on Mercury when having two items with the same title inside featured. It goes to this check: extensions/wikia/CuratedContent/CuratedContentValidator.class.php:56 please fix.

**Comment by bognix** [comment]
@vforge please fix: http://qa-s2.wikia-prod:8080/job/unitTests_php_branches/18863/testReport/junit/(root)/CuratedContentValidatorTest__testValidation/testValidation_with_data_set__10/

**Comment by vforge** [comment]
@bognix fixed in http://qa-s2.wikia-prod:8080/job/unitTests_php_branches/18867/

File: extensions/wikia/CuratedContent/CuratedContentValidator.class.php [hunk 3]
```diff
@@ -112,10 +115,7 @@ public function validateSection( $section ) {
 
 		if ( empty( $section['featured'] ) && !empty( $section['title'] ) ) {
 			$this->validateImage( $section );
-		}
-
-		if ( strlen( $section['title'] ) ) {
-			$this->titles[] = $section['title'];
+			$this->addLabelForValidation( $section['title'], $section['title'] );
```
Comment (bognix) [inline extensions/wikia/CuratedContent/CuratedContentValidator.class.php]: Add comment that label for section is the same as title

File: extensions/wikia/CuratedContent/CuratedContentValidator.class.php [hunk 4]
```diff
@@ -160,7 +160,15 @@ public function validateItem( $item ) {
...
 
-		$this->titles[] = $item['title'];
+		$this->addLabelForValidation( $item['label'], $item['title'] );
+	}
+
+	private function addLabelForValidation( $label, $title ) {
+		if ( array_key_exists( $label, $this->existingLabels ) ) {
+			$this->existingLabels[$label][] = $title;
+		}  else {
+			$this->existingLabels[$label] = [ $title ];
```
Comment (bognix) [inline extensions/wikia/CuratedContent/CuratedContentValidator.class.php]: please simplify.
This $this->existingLabels[$label][] = $title; is enough no ifs ;)

**Comment by bognix** [comment]
LGTM

**Comment by tomasznapieralski** [comment]
LGTM

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js [hunk 5]
```diff
@@ -327,56 +333,32 @@ require(['wikia.window', 'jquery', 'wikia.nirvana', 'wikia.tracker', 'JSMessages
...
-						}
-						if ($itemWithError) {
-							$itemWithError.addError(reasonMessage);
-							return false;
-						}
-					}
-					return true;
-				});
+			function addErrorToSection($selector, errReason, reasonMessage) {
+				switch (errReason) {
```
Comment (tomasznapieralski) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: what do you think about converting `switch` to `if..else`, cause you have one option and default?

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js [hunk 6]
```diff
@@ -408,17 +390,43 @@ require(['wikia.window', 'jquery', 'wikia.nirvana', 'wikia.tracker', 'JSMessages
...
+							[].forEach.call(data.error, function(error) {
+								// error := { target: <label>, type: [ item,section,featured], reason: error }
+								var errLabel = error.target,
+									errType = error.type,
+									errReason = error.reason,
 									reasonMessage = gerErrorMessageFromErrReason(errReason);
 
-								checkItemsForErrors($items, errTitle, errReason, reasonMessage);
-								checkSectionsForErrors($sections, errTitle, errReason, reasonMessage);
+								if (errType === 'item') {
```
Comment (tomasznapieralski) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: what do you think about converting `if..else` to `switch`?

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js (hunk 5 above)
Comment (vforge) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: What would be benefit from that? I wanted it to be as similar to `addErrorToItem` as possible - do you think of any reason why I should change it so much?

Comment (tomasznapieralski) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: only readability

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js (hunk 6 above)
Comment (vforge) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: What do you think about converting `if...else` to functions to increase readability (see indentation)

Comment (tomasznapieralski) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: function will be nice

**Comment by tomasznapieralski** [comment]
LGTM

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js [hunk 7]
```diff
@@ -327,55 +333,49 @@ require(['wikia.window', 'jquery', 'wikia.nirvana', 'wikia.tracker', 'JSMessages
...
-							// intended fall
-							case 'tooLongLabel':
-								$itemWithError = $(this);
-								break;
-						}
-						if ($itemWithError) {
-							$itemWithError.addError(reasonMessage);
-							return false;
+			function addErrorToSection($selector, reason, message) {
+				if (reason === 'imageMissing') {
```
Comment (hakubo) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]:
``` js
$selector.find(
  (reason === 'imageMissing') ? '.image' :  '.section-input'
).addError(message);
```

File: extensions/wikia/CuratedContent/CuratedContentValidator.class.php [hunk 8]
```diff
@@ -100,28 +124,26 @@ public function validateSection( $section ) {
...
-			$this->titles[] = $section['title'];
+			// label for section is the same as it's title
+			$this->existingSectionLabels[] = $section['title'];
 		}
 	}
 
 	public function validateItemType( $item ) {
-		if ( $item['type'] !== CuratedContentHelper::STR_CATEGORY ) {
-			$this->error( $item, self::ERR_NO_CATEGORY_IN_TAG );
+		if ( !in_array( $item['type'], [ CuratedContentHelper::STR_CATEGORY ] ) ) {
```
Comment (rogatty) [inline extensions/wikia/CuratedContent/CuratedContentValidator.class.php]: What's the difference?

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js [hunk 9]
```diff
@@ -327,55 +333,49 @@ require(['wikia.window', 'jquery', 'wikia.nirvana', 'wikia.tracker', 'JSMessages
...
-						}
-						if ($itemWithError) {
-							$itemWithError.addError(reasonMessage);
-							return false;
-						}
-					}
-					return true;
-				});
+			function addErrorToItem($selector, reason, message) {
+				if (reason === 'imageMissing') {
```
Comment (hakubo) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]:
``` js

function getSelector(list, reason) {
  return list[reason] || list.default;
}

$selector.find(
  getSelector({
    imageMissing: '.image',
    emptyLabel: '.name',
    duplicatedLabel: '.name',
    tooLongLabel: '.name',
    default: '.item-input'
  }, reason)
).addError(message);
```

File: extensions/wikia/CuratedContent/CuratedContentValidator.class.php [hunk 10]
```diff
@@ -133,41 +155,46 @@ public function validateItemsTypes( $section ) {
...
 		}
 
-		if ( $item['type'] === CuratedContentHelper::STR_VIDEO ) {
-			if ( empty( $item['video_info'] ) ) {
-				$this->error( $item, self::ERR_VIDEO_WITHOUT_INFO );
-			} elseif ( !self::isSupportedProvider( $item['video_info']['provider'] ) ) {
-				$this->error( $item, self::ERR_VIDEO_NOT_SUPPORTED );
-			}
+		if ( self::needsArticleId( $item['type'] ) && empty( $item['article_id'] ) ) {
+			$this->error( $item['label'], $isFeatured ? 'featured' : 'item', self::ERR_ARTICLE_NOT_FOUND );
```
Comment (rogatty) [inline extensions/wikia/CuratedContent/CuratedContentValidator.class.php]: How about extracting this ternary op to a variable? It's used 4 times.

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js (hunk 7 above)
Comment (hakubo) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: or use 

``` js

$selector.find(
  getSelector({
    imageMissing: '.image',
    default: '.section-input'
  }, reason)
).addError(message);
```

**Comment by hakubo** [comment]
One idea for abstraction otherwise  LGTM

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js [hunk 11]
```diff
@@ -408,17 +408,18 @@ require(['wikia.window', 'jquery', 'wikia.nirvana', 'wikia.tracker', 'JSMessages
...
+							$items = $ul.find('.section:not(.featured)').first().nextAll(),
+							$sections = $ul.find('.section:not(.featured)');
 
-							[].forEach.call(data.error, function(err) {
-								var errTitle = err.title,
-									errReason = err.reason,
-									reasonMessage = gerErrorMessageFromErrReason(errReason);
+						if (data.error) {
+							[].forEach.call(data.error, function(error) {
+								// error := { target: <label>, type: [ item,section,featured], reason: error }
```
Comment (rogatty) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: Nice!

**Comment by Warkot** [comment]
LGTM

**Comment by rogatty** [comment]
LGTM

File: extensions/wikia/CuratedContent/CuratedContentHelper.class.php [hunk 12]
```diff
@@ -54,12 +53,7 @@ public function fillItemInfo( &$item ) {
 
 				case self::STR_CATEGORY:
 					$category = Category::newFromTitle( $title );
-					if ( $category instanceof Category && $category->getID() ) {
-						$count = $category->getPageCount();
-						if ( empty( $count ) ) {
-							$item['type'] = self::STR_EMPTY_CATEGORY;
-						}
-					} else {
+					if ( !( $category instanceof Category && $category->getID() ) ) {
```
Comment (bognix) [inline extensions/wikia/CuratedContent/CuratedContentHelper.class.php:107]: So you are setting 'type' to null for empty categories? If so, it won't work.
Take this category for instance: http://concf.wikia.com/wiki/Category:Lorem. It's wgArticleId is 0 and it is an instance of a Category class and it has some items - so it's valid candidate for Curated Content but you would set 'type' to null

Comment (vforge) [inline extensions/wikia/CuratedContent/CuratedContentHelper.class.php:107]: `getID()` is not returning `wgArticleId` - it's returning `cat_id`:

```
select * from category where cat_title = 'Lorem';
+--------+-----------+-----------+-------------+-----------+------------+
| cat_id | cat_title | cat_pages | cat_subcats | cat_files | cat_hidden |
+--------+-----------+-----------+-------------+-----------+------------+
|     15 | Lorem     |         5 |           0 |         1 |          0 |
+--------+-----------+-----------+-------------+-----------+------------+
1 row in set (0.00 sec)
```

Comment (bognix) [inline extensions/wikia/CuratedContent/CuratedContentHelper.class.php:107]: perfect! thanks for checking!

**Comment by bognix** [comment]
LGTM

File: extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js [hunk 13]
```diff
@@ -408,17 +410,18 @@ require(['wikia.window', 'jquery', 'wikia.nirvana', 'wikia.tracker', 'JSMessages
...
+							$sections = $ul.find('.section:not(.featured)');
 
-							[].forEach.call(data.error, function(err) {
-								var errTitle = err.title,
-									errReason = err.reason,
-									reasonMessage = gerErrorMessageFromErrReason(errReason);
+						if (data.error) {
+							[].forEach.call(data.error, function(error) {
+								// error := { target: <label>, type: [item,section,featured], reason: <error> }
+								var message = gerErrorMessageFromErrReason(errReason);
```
Comment (rogatty) [inline extensions/wikia/CuratedContent/js/CuratedContentManagmentTool.js]: @vforge `errReason` is undefined

**Comment by hakubo** [comment]
aaaaand LGTM :)

**Comment by bognix** [comment]
LGTM
